### PR TITLE
[Store] Decouple business decisions from storage metrics

### PR DESCRIPTION
**Pull request by Hubert-Zhu**
## Description

This refactor makes storage metrics a projection of authoritative Store domain state instead of an input to business decisions:

- Add DRAM/NoF usage snapshots derived from mounted allocators (with shared CXL allocator deduplication).
- Keep O(1) aggregate DRAM/NoF reads for eviction and promotion via a simple domain usage tracker (metric-gauge style atomics), while retaining O(n) per-segment snapshots for observability.
- Refresh Prometheus DRAM/NoF gauges from tiered snapshots on admin metric reads and periodic reports; clear stale HA leader projections on standby.
- Remove business-oriented DRAM/NoF ratio APIs from `MasterMetricManager`.
- Cover metric reset, snapshot restore/reset, allocator replacement, standby/HA transition, shared CXL, NoF, lock isolation, and Prometheus projection behavior.

Implementation is split into reviewable slices:

1. Decouple DRAM eviction and promotion decisions from metrics.
2. Project storage metrics from domain snapshots, with API cleanup and resilience tests.
3. Make business-path domain usage reads O(1) with a lightweight tracker.
4. Harden allocator lifetime, CXL publication, and HA metric handoff.

## Module

- [ ] Transfer Engine (`mooncake-transfer-engine`)
- [x] Mooncake Store (`mooncake-store`)
- [ ] Mooncake EP (`mooncake-ep`)
- [ ] Mooncake PG (`mooncake-pg`)
- [ ] Integration (`mooncake-integration`)
- [ ] P2P Store (`mooncake-p2p-store`)
- [ ] Python Wheel (`mooncake-wheel`)
- [ ] Common (`mooncake-common`)
- [ ] Mooncake RL (`mooncake-rl`)
- [ ] CI/CD
- [x] Docs
- [ ] Other

## Type of Change

- [x] Bug fix
- [ ] New feature
- [x] Refactor
- [ ] Breaking change
- [x] Documentation update
- [x] Performance improvement
- [ ] Other

## How Has This Been Tested?

**Test commands:**
```bash
./scripts/code_format.sh -b origin/main
pre-commit run --files <touched files>
# unit coverage added/updated in:
#   segment_test, master_metrics_test, promotion_on_hit_test
```

**Test results:**
- [ ] Unit tests pass
- [ ] Integration tests pass (if applicable)
- [ ] Manual testing done (describe below)

Local formatting/pre-commit were run on touched files during preparation. Full compile/unit validation is expected from Linux PR CI.

## Checklist

- [ ] I have performed a self-review of my own code
- [x] I have formatted my code using `./scripts/code_format.sh`
- [ ] I have run `pre-commit run --all-files` and all hooks pass
- [x] I have updated the documentation (if applicable)
- [x] I have added tests to prove my changes are effective
- [x] For changes >500 LOC: I have filed an RFC issue (#3158)

## AI Assistance Disclosure

- [ ] No AI tools were used
- [x] AI tools were used (specify below)

Cursor and Codex

### CODE REVIEW

**Comment by Hubert-Zhu** [comment]
I have not full self-review all the test cases, but the feature code is self-reviewed line by line. This is my first big PR in this repo. Please forgive me if I miss any edge cases.

**Comment by Aionw** [comment]
@Hubert-Zhu While reviewing the new aggregate usage tracker lifecycle, I found a concurrency issue around explicit detachment.

`UnmountSegment()`, `Remove()`, and `BatchEvict()` all take `snapshot_mutex_` in shared mode, so they can run concurrently. During unmount, `PrepareUnmountSegment()` holds `segment_mutex_` and calls `allocator->DetachUsageTracker()`, which subtracts the allocator's current contribution and resets `usage_tracker_`. At the same time, an `AllocatedBuffer` destroyed by Remove/eviction can successfully `weak_ptr::lock()` the same allocator and enter `deallocate() -> RecordDeallocation()` without taking `segment_mutex_`.

That creates two problems:

1. `usage_tracker_` is read and reset concurrently, which is a data race on the shared_ptr object.
2. Depending on the interleaving, detach can subtract the allocator's full usage while deallocation also subtracts its delta, potentially underflowing the aggregate `size_t`. Since eviction and promotion now read this tracker directly, the corruption can persist and affect business decisions.

I do not think we should put a mutex/shared_mutex on every allocate/deallocate just to protect this rare lifecycle transition. A zero-hot-path-overhead design would be:

- Attach exactly once before an allocator is published; keep its registration immutable while the allocator is live.
- Remove the public/general-purpose `DetachUsageTracker()` path.
- On normal unmount, remove the allocator from `AllocatorManager` and release the mounted shared_ptr, but do not explicitly detach.
- An in-flight deallocation already holds a local allocator shared_ptr, so the allocator cannot be destroyed until that operation finishes. Let the allocator destructor RAII-detach the registration when the last shared_ptr is released.
- Model the per-allocator attachment as an allocator-owned `std::unique_ptr<StorageUsageRegistration>`; the registration can retain the shared aggregate `StorageUsageTracker`.
- Handle allocator replacement/CXL replacement as private, explicitly quiescent ownership-transfer operations: release the old allocator before publishing/attaching the new one, or transfer the registration internally. They should not require a generally callable detach API.

The aggregate may retain an unmounted allocator for the short period needed for already-running operations to drain, which seems compatible with the documented best-effort topology semantics and avoids adding synchronization to the allocation hot path.

It would also be valuable to add a forced-interleaving or TSAN test covering concurrent `AllocatedBuffer` destruction and `PrepareUnmountSegment()`.

**Comment by Hubert-Zhu** [comment]
Make sense. I will fix it.

File: mooncake-store/src/master_metric_manager.cpp [hunk 1]
```diff
@@ -761,13 +743,42 @@ int64_t MasterMetricManager::get_total_nof_capacity() {
...
+        for (const auto& [segment_name, usage] : tier.segments) {
+            current_segments.insert(segment_name);
+            allocated_by_segment.update({segment_name},
+                                        static_cast<int64_t>(usage.used_bytes));
+            capacity_by_segment.update(
+                {segment_name}, static_cast<int64_t>(usage.capacity_bytes));
+        }
+        for (const auto& segment_name : projected_segments) {
+            if (!current_segments.contains(segment_name)) {
+                allocated_by_segment.update({segment_name}, 0);
```
Comment (Aionw) [inline mooncake-store/src/master_metric_manager.cpp]: [P2] Remove labels that disappear from the authoritative snapshot

`dynamic_gauge::update(..., 0)` keeps or recreates the label. Immediately afterward, `projected_segments` is replaced with `current_segments`, so the next projection no longer knows about the missing segment and can never remove that zero-valued series. Repeated mount/unmount cycles therefore accumulate stale Prometheus label sets. Please call `remove_label_value({{"segment", segment_name}})` on both per-segment gauges for entries absent from the new snapshot, and add a test asserting that serialized metrics no longer contain the removed segment.

File: mooncake-store/src/allocator.cpp [hunk 2]
```diff
@@ -11,6 +11,34 @@
...
+
+void BufferAllocatorBase::RecordAllocation(size_t bytes) noexcept {
+    cur_size_.fetch_add(bytes, std::memory_order_relaxed);
+    if (usage_registration_) {
+        usage_registration_->AddUsedBytes(bytes);
+    }
+}
+
+void BufferAllocatorBase::RecordDeallocation(size_t bytes) noexcept {
+    if (auto hook = record_deallocation_hook_.load(std::memory_order_acquire)) {
```
Comment (Aionw) [inline mooncake-store/src/allocator.cpp]: [P2] Keep test coordination out of the deallocation hot path

This testing hook adds an acquire atomic load and branch to every production deallocation, exposes a global mutable test API on `BufferAllocatorBase`, and makes the test rely on a process-global function pointer and static state. The race can be forced without touching production code: define a test-only `BlockingAllocator` or subclass `OffsetBufferAllocator`, override the virtual `deallocate()`, signal an entered semaphore, wait on a resume semaphore, and then delegate to the real deallocation or `RecordDeallocation`. That proves the local `shared_ptr` keeps the allocator and registration alive while unmount drops its reference, with zero production-path overhead. Please remove the setter, static atomic, and hook branch.

**Review by Aionw — APPROVED** [review]
LGTM

**Review by ykwd — APPROVED** [review]
